internal/ignition/ignition.go

```diff
@@ -1078,18 +1078,27 @@ func uploadToS3(ctx context.Context, workDir string, cluster *common.Cluster, s3
 	return nil
 }
 
+// ParseToLatest takes the Ignition config and tries to parse it as v3.1 and if that fails,
+// as v3.2. This is in order to support the latest possible Ignition as well as to preserve
+// backwards-compatibility with OCP 4.6 that supports only Ignition up to v3.1
 func ParseToLatest(content []byte) (*config_latest_types.Config, error) {
```

Comment:
I'm not sure why this is required?
Currently, 3.2 is the default (unless we fail to parse).
In case the config isn't compatible with 3.2 we fall back to 3.1.
This change reverse the order so that now 3.1 will be the default

Reply:
You are correct, I'm just not sure why you want to change the current behavior?
According to previous commits to this file:

> The strategy will always be to parse up to the latest supported
    version then manipulate types that will not change in future minor
    ignition versions (File, Config, Unit, etc).
    
Note that 4.6 base template fails to parse as 3.2 (otherwise the installation will fail) so it get's the fall back.
 The common case and (future releases) would rather get the latest version we support

Reply:
The part that I don't understand in the current behaviour (and in fact, what I don't understand in the commit message) is how Assisted when calling this function knows whether it works with base template for 4.6 or later. The template for day2 worker is being built in runtime, as in 

```
	"ignition": {
	  "version": "3.1.0",
	  "config": {
		"merge": [{
		  "source": insert-mcs-location-here.test:12345
		}]
	  }
	}
```
and then Ignition service inside RHCOS calls MCS and merges the manifest. The problem I had initially was that bumping this version to 3.2.0 was feeding RHCOS with unsupported ignition.

I apply equivalent reasoning to the bootstrap ignition. We have a base template (note that I don't modify it in the PR)
```
  "ignition": {
    "version": "3.1.0"
  },
  "passwd": {
    "users": [
      {{.userSshKey}}
    ]
  },
[...]
```
that we must not bump to 3.2.0 in here, because if we do so, we will feed OCP 4.6 with 3.2.0 and that fails.

So based on those 2 templates above, I don't fully understand the logic of 

>Currently, 3.2 is the default (unless we fail to parse).
In case the config isn't compatible with 3.2 we fall back to 3.1.

because our base templates are defined as 3.1.0 and we artificially bump them to 3.2.0 by changing the string of version field. Even though our unit tests currently show that stuff seems to work, I can't answer the following question - **which piece of code is responsible for generating 3.1.0 for 4.6 and 3.2.0 for later if none of those functions are parametrized by cluster version?**

After all, currently stuff works for day1 but day2 is not implemented. At the same time I'm not sure we all understand the current code and logic which determines which ignition version is generated when, do we? Maybe the whole action of "3.1.0 by default unless specifically requested otherwise" should be treated as a cleanup?

Reply:
> which piece of code is responsible for generating 3.1.0 for 4.6 and 3.2.0 for later if none of those functions are parametrized by cluster version?

`openshift-install` cli.
When we generate the manifests with the chosen cli version we get whatever ignition comes out of that invocation.
The idea when this was written was to avoid us ever having to understand the mapping between ocp and ignition version. We effectively just pass this responsibility on to the user.

For discovery ignition it's simpler as we're defining the template version and that will work with any rhcos version so we will only need to bump it if and when we need newer features. Overrides must then just match our template version.

I think this is more confusing because the cluster generates (and expects) one version, we have a template with a version, and we have user overrides and we need all of that to work together. Does that sound right?

Reply:
Ok, after all that I'm keeping the current behaviour. We will generate 3.2.0 for as long as we can and only move to 3.1.0 if needed.